croncheck.php

```diff
@@ -121,8 +121,45 @@
     header('Expires: Tue, 04 Sep 2012 05:32:29 GMT');
 }
 
+if (isset($CFG->adminsetuppending)) {
+    send_critical("Admin setup pending, please set up admin account");
+}
+
 if (moodle_needs_upgrading()) {
-    send_critical("Moodle upgrade pending, cron execution suspended");
+    // Check if an upgrade is running.
+    if (!isset($CFG->upgraderunning) || $CFG->upgraderunning < time()) {
+        $upgraderunning = get_config(null, 'upgraderunning');
+    } else {
+        $upgraderunning = $CFG->upgraderunning;
+    }
+
+    $initialinstall = during_initial_install();
+
+    if ($upgraderunning && $upgraderunning >= time()) {
+        // Before the expected finish time.
+        if (!empty($initialinstall)) {
+            send_critical("Moodle initial installation is running, cron execution suspended");
+        } else {
+            send_critical("Moodle upgrade is running, cron execution suspended");
+        }
+    } else if ($upgraderunning && $upgraderunning < time()) {
+        $difference = time() - $upgraderunning;
+        /*
+         * After the expected finish time (timeout or other interruption)
+         * The "core_shutdown_manager::register_function('upgrade_finished_handler');" already handle these cases
+         * and unset config 'upgraderunning'
+         * The below critical ones can happen if core_shutdown_manager fails to run the handler function.
+         */
+        if (!empty($initialinstall)) {
+            send_critical("Moodle initial installation is running and
+            expected finish time has been elapsed for $difference seconds, cron execution suspended");
```

Comment:
Nagios messages should be < 80 chars, they can be multi line. I think just reword these to be more susinct like

"Moodle initial installation is running and overdue by 12 mins"

Are you able to say how long it has been running for? eg

"Moodle initial installation has been running for 10 mins and overdue by 12 mins"

It is not too important to mention cron here

Reply:
I have changed the wordings.
I cannot find way to calculate the duration.

There is a class call "core_upgrade_time" which is used to calculate duration for each plugin, but I cannot use it as it reset after each plugin upgrade